pennylane_lightning_gpu/src/algorithms/AdjointDiffGPU.hpp

```diff
@@ -502,6 +504,203 @@ template <typename T> class HamiltonianGPU final : public ObservableGPU<T> {
+
+        cudaDataType_t data_type;
+        cusparseIndexType_t compute_type;
+
+        if constexpr (std::is_same_v<CFP_t, cuDoubleComplex> ||
+                      std::is_same_v<CFP_t, double2>) {
+            data_type = CUDA_C_64F;
+            compute_type = CUSPARSE_INDEX_64I;
+        } else {
+            data_type = CUDA_C_32F;
+            compute_type = CUSPARSE_INDEX_32I;
+        }
+
+        // CUSPARSE APIs
+        cusparseHandle_t handle = nullptr;
+        cusparseSpMatDescr_t mat;
+        cusparseDnVecDescr_t vecX, vecY;
+
+        size_t bufferSize = 0;
+
+        PL_CUSPARSE_IS_SUCCESS(cusparseCreate(&handle))
+
+        // Create sparse matrix A in CSR format
+        PL_CUSPARSE_IS_SUCCESS(cusparseCreateCsr(
+            /* cusparseSpMatDescr_t* */ &mat,
+            /* int64_t */ num_rows,
+            /* int64_t */ num_cols,
+            /* int64_t */ nnz,
+            /* void* */ d_csrOffsets.getData(),
+            /* void* */ d_columns.getData(),
+            /* void* */ d_values.getData(),
+            /* cusparseIndexType_t */ compute_type,
+            /* cusparseIndexType_t */ compute_type,
+            /* cusparseIndexBase_t */ CUSPARSE_INDEX_BASE_ZERO,
+            /* cudaDataType */ data_type))
+
+        // Create dense vector X
+        PL_CUSPARSE_IS_SUCCESS(cusparseCreateDnVec(
+            /* cusparseDnVecDescr_t* */ &vecX,
+            /* int64_t */ num_cols,
+            /* void* */ sv.getData(),
+            /* cudaDataType */ data_type))
+
+        // Create dense vector y
+        PL_CUSPARSE_IS_SUCCESS(cusparseCreateDnVec(
+            /* cusparseDnVecDescr_t* */ &vecY,
+            /* int64_t */ num_rows,
+            /* void* */ sv.getData(),
+            /* cudaDataType */ data_type))
+
+        // allocate an external buffer if needed
+        PL_CUSPARSE_IS_SUCCESS(cusparseSpMV_bufferSize(
+            /* cusparseHandle_t */ handle,
+            /* cusparseOperation_t */ CUSPARSE_OPERATION_NON_TRANSPOSE,
+            /* const void* */ &alpha,
+            /* cusparseSpMatDescr_t */ mat,
+            /* cusparseDnVecDescr_t */ vecX,
+            /* const void* */ &beta,
+            /* cusparseDnVecDescr_t */ vecY,
+            /* cudaDataType */ data_type,
+            /* cusparseSpMVAlg_t */ CUSPARSE_SPMV_CSR_ALG1,
+            /* size_t* */ &bufferSize)) // Can also use CUSPARSE_MV_ALG_DEFAULT
+
+        DataBuffer<void, int> dBuffer{bufferSize, device_id, stream_id, true};
+
+        // execute SpMV
+        PL_CUSPARSE_IS_SUCCESS(cusparseSpMV(
+            /* cusparseHandle_t */ handle,
+            /* cusparseOperation_t */ CUSPARSE_OPERATION_NON_TRANSPOSE,
+            /* const void* */ &alpha,
+            /* cusparseSpMatDescr_t */ mat,
+            /* cusparseDnVecDescr_t */ vecX,
+            /* const void* */ &beta,
+            /* cusparseDnVecDescr_t */ vecY,
+            /* cudaDataType */ data_type,
+            /* cusparseSpMVAlg_t */ CUSPARSE_SPMV_CSR_ALG1,
```

Comment:
I think we need to move the comment  // Can also use // CUSPARSE_MV_ALG_DEFAULT in line 684 to line 683.

Reply:
Done